### PR TITLE
Switch to using zod for validation for patch endpoint.

### PR DESCRIPTION
**Pull request by Victor-M-Giraldo**
- Related #1385

### CODE REVIEW

**Comment by Victor-M-Giraldo** [comment]
I think this is in the right direction. Would just like some feedback before I continue

File: common/models/zod-schemas.ts [hunk 1]
```diff
@@ -55,3 +56,44 @@ export const OttApiRequestAccountRecoveryVerifySchema = z.object({
...
+		grants: z.array(
+			z.tuple([
+				z.object({
+					role: z.nativeEnum(Role),
+				}),
+				z.object({
+					grantMask: z.number(),
+				}),
+			])
+		),
```
Comment (dyc3) [inline common/models/zod-schemas.ts:92]: This should go in `RoomSettingsSchema`

File: common/models/zod-schemas.ts [hunk 2]
```diff
@@ -55,3 +56,44 @@ export const OttApiRequestAccountRecoveryVerifySchema = z.object({
...
+		...RoomSettingsSchema.shape,
+	}),
+	z.object({
+		claim: z.boolean().optional(),
+		...RoomSettingsSchema.shape,
```
Comment (dyc3) [inline common/models/zod-schemas.ts:97]: and `RoomSettingsSchema` should be excluded from here.

Comment (Victor-M-Giraldo) [reply common/models/zod-schemas.ts]: To clarify, did you mean they should be excluded from both? Or is how it is now fine?

Still needs to be a union type right? So union type between the claim and the rest of the room settings fields?

File: common/models/zod-schemas.ts [hunk 3]
```diff
@@ -55,3 +56,40 @@ export const OttApiRequestAccountRecoveryVerifySchema = z.object({
...
+	grants: z.array(GrantSchema).optional(),
+	autoSkipSegmentCategories: z.array(CategorySchema).optional(),
+	restoreQueueBehavior: z.nativeEnum(BehaviorOption).optional(),
+	enableVoteSkip: z.boolean().optional(),
+});
+
+const ClaimSchema = z.object({
+	claim: z.boolean().optional(),
+	...RoomSettingsSchema.shape,
+});
```
Comment (dyc3) [inline common/models/zod-schemas.ts:93]: I meant just the object with `claim`. It shouldn't include room settings.
```suggestion
const ClaimSchema = z.object({
	claim: z.boolean().optional(),
});
```

File: common/models/zod-schemas.ts [hunk 4]
```diff
@@ -55,3 +56,40 @@ export const OttApiRequestAccountRecoveryVerifySchema = z.object({
...
+	"preview",
+]);
+
+const GrantSchema = z.tuple([
+	z.object({
+		role: z.nativeEnum(Role),
+	}),
+	z.object({
+		grantMask: z.number(),
+	}),
```
Comment (dyc3) [inline common/models/zod-schemas.ts:76]: This doesn't look right. Its a tuple of 2 objects, not of the enum and the mask

**Comment by Victor-M-Giraldo** [comment]
Just giving an update on this one. Still working on it, I'm aware I need to narrow the type down but I'm also encountering an issue after I do that. It's a WIP though.

**Comment by dyc3** [comment]
IIRC the only endpoint that uses query params is `/api/data/previewAdd`, which is in `data.ts`

**Comment by dyc3** [comment]
As discussed offline:

for handling `grants`, create a new object from `body`

probably something like
```
const foo = {
...body,
grants: new Grants(body.grants)
}
```

be careful to not add it if it doesn't exist on the body

**Comment by Victor-M-Giraldo** [comment]
I know you said not to do `else if ('visibility' in body)` but if I just do `else` I get type errors on the grants. Also, I'm getting type errors when trying to update the settings to just be the body and it says the Grants is missing some properties.

File: server/api/room.ts [hunk 5]
```diff
@@ -184,76 +184,63 @@ const getRoom: RequestHandler<{ name: string }, OttApiResponseGetRoom, unknown>
...
+	} else if ('visibility' in body) {
+		const newBody = {
+			...body,
+			grants: new Grants(body.grants),
+		}
+
 		const roomRequest: ApplySettingsRequest = {
 			type: RoomRequestType.ApplySettingsRequest,
-			settings: req.body,
+			settings: body,
```
Comment (dyc3) [inline server/api/room.ts:239]: should be
```suggestion
			settings: newBody,
```
or
```suggestion
			settings: {
				...body,
				grants: new Grants(body.grants),
			},
```

File: server/api/room.ts [hunk 6]
```diff
@@ -184,76 +184,63 @@ const getRoom: RequestHandler<{ name: string }, OttApiResponseGetRoom, unknown>
...
 		};
 
 		await room.processUnauthorizedRequest(roomRequest, { token: req.token });
+
```
Comment (dyc3) [inline server/api/room.ts:243]: nit: stray newline

**Review by dyc3 — CHANGES_REQUESTED** [review]

**Comment by Victor-M-Giraldo** [comment]
Still need to fix tests.

**Review by dyc3 — APPROVED** [review]

**Pull request merged**